orderbook/src/orderbook.rs

```diff
@@ -57,17 +60,25 @@ impl Orderbook {
         event_updater: EventUpdater,
         balance_fetcher: Box<dyn BalanceFetching>,
         fee_validator: Arc<MinFeeCalculator>,
+        deny_tokens: HashSet<H160>,
     ) -> Self {
         Self {
             domain_separator,
             database,
             event_updater: Mutex::new(event_updater),
             balance_fetcher,
             fee_validator,
+            deny_tokens,
         }
     }
 
     pub async fn add_order(&self, order: OrderCreation) -> Result<AddOrderResult> {
+        if self.deny_tokens.contains(&order.buy_token) {
+            return Ok(AddOrderResult::BuyTokenDenied(order.buy_token));
+        }
```

Comment:
If you put `contains_token_from` from before on `OrderCreation` instead of `Order` you can reuse it here but I guess then we can't differentiate between sell token being the error or buy token. I'd also be fine with having only one error instead of two. No need to change anything, just mentioning.

Reply:
I like the idea of reuse, but I think its nice to tell the user exactly why there order was rejected if we can. Open to influence if anyone has strong opinion, but would otherwise leave it as is.